concordium-std/src/bump_alloc.rs

```diff
@@ -0,0 +1,241 @@
+    /// specifies which memory address the heap starts at. Memory addresses
+    /// prior to that are used for the data and the stack.
+    ///
+    /// To get the actual memory location as a `usize`, do the following:
+    ///
+    /// ```
+    /// let heap_base = unsafe { &__heap_base as *const _ as usize }; 
+    /// ```
+    static __heap_base: u8;
+}
+
+/// This is an invalid implementation of [`Sync`].
+/// The [`BumpAllocator`] cannot be safely used from multiple threads, but that
+/// is OK since it won't be in Wasm. This [`Sync`] implementation is required
+/// for defining it as the global allocator.
+unsafe impl Sync for BumpAllocator {}
+
+/// A bump allocator for Wasm.
+///
+/// See the module documentation for more details.
+pub struct BumpAllocator {
+    /// The pointer to the next memory to hand out.
+    next:        UnsafeCell<usize>,
+    /// The start of the heap.
+    /// This value is used for resetting `next` if all memory chunks are
+    /// deallocated and `allocations` thus becomes 0.
+    /// The actual location of the heap cannot be known at compile time and
+    /// the initial value is thus 0. It is set to `__heap_base` value on the
+    /// first allocation.
+    heap_start:  UnsafeCell<usize>,
+    /// The end of the heap. Cannot be known at compile time and thus the
+    /// initial value is 0. It is updated on the first allocation.
+    heap_end:    UnsafeCell<usize>,
+    /// The number of active allocations. Used for resetting `next` to
+    /// `heap_start` if there are no more active allocations.
+    allocations: UnsafeCell<usize>,
+    /// Stores the last address given out, and allows for a small improvement
+    /// over the always-increasing bump allocator. Namely that if an item
+    /// `a` is allocated and deallocated without any allocations in between,
+    /// then the memory of `a` is reused for the next allocation.
+    last_alloc:  UnsafeCell<usize>,
+}
+
+impl BumpAllocator {
+    /// Create a new [`BumpAllocator`].
+    ///
+    /// This must be a `const` method for it to be used as a global allocator,
+    /// where it is called staticly. Since we cannot know the start and end
+    /// locations of the heap on compile time, several values are initialized to
+    /// a dummy value `0` and updated appropriately during the first allocation.
+    ///
+    /// # Safety
+    /// - Can only be used in single-threaded environments. The [`Sync`]
+    ///   implementation is only invalid and is only there to support setting
+    ///   this as the global allocator.
+    pub const unsafe fn new() -> Self {
+        Self {
+            // Initialized to the dummy value `0`.
+            next:        UnsafeCell::new(0),
+            // Initialized to the dummy value `0`.
+            heap_start:  UnsafeCell::new(0),
+            // Initialized to the dummy value `0`, which is checked during first initialization.
+            heap_end:    UnsafeCell::new(0),
+            allocations: UnsafeCell::new(0),
```

Comment:
Could you document what this field is for, and why it is needed?

Reply:
It is documented on the type. But I also added a line here.